core/src/main/scala/chisel3/experimental/dataview/package.scala

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.experimental
+
+import chisel3._
+import chisel3.internal._
+import chisel3.internal.sourceinfo.SourceInfo
+
+import scala.annotation.{implicitNotFound, tailrec}
+import scala.collection.mutable
+import scala.collection.immutable.LazyList // Needed for 2.12 alias
+
+package object dataview {
+  case class InvalidViewException(message: String) extends ChiselException(message)
+
+  // TODO should we move the `DataProduct` requirement to `viewAs`
```

Comment:
Why?

Reply:
Because then you could call `.viewAs` on anything and it would let you know that you need a `DataProduct` rather than it just saying "method not found". Will play with this and either do that move if it makes sense or delete this comment.

Reply:
Given the following code:
```scala
case class Foo(val x: UInt)
class MyModule extends Module {
  val x = Foo(3.U)
  x.viewAs[UInt]
  val in = IO(Input(UInt(8.W)))
  val out = IO(Output(UInt(8.W)))
  out := in
}
```
Without that TODO being done, you get the error:
```
[error] .../Example.scala:8:5: value viewAs is not a member of Foo
[error]   x.viewAs[UInt]
[error]     ^            
```
(which is the same error you get if you forget `import chisel3.experimental.dataview._`)

If I move `DataProduct` to `viewAs` instead of on `DataViewable`, you get:
```
[error] .../File.scala:8:11: Could not find implicit value for DataProduct[Foo].
[error] Please see https://www.chisel-lang.org/chisel3/docs/explanations/dataview#dataproduct
[error]   x.viewAs[UInt]
[error]           ^
```

Reply:
So I went ahead and moved it.